File: legacy/badrcptto.pl
#!/usr/bin/perl

use strict;
use warnings;

my $recipient = $ENV{'SMTPRCPTTO'};

# Relay- and trustclients pass right trough
# If we have no configfile we give up
if ( (defined($ENV{'RELAYCLIENT'})) || 
     (defined($ENV{'TRUSTCLIENT'})) ||
     (not(-r "/var/qmail/control/badrcptto")) ) {
  print "\n";
  exit(0);
}

# Check the list of bad recepients
open(BADRCPTTO,  "/var/qmail/control/badrcptto");
while(<BADRCPTTO>) {
  chomp($_);
  if ($recipient eq $_) {
    print STDERR getppid()." Found $recipient in badrcptto list !\n";
    print "E550 This address no longer accepts mail [$recipient]\n";
    close(BADRCPTTO);
    exit(0);
  }
}
close(BADRCPTTO);

print "\n";
exit(0);

File: badrcptto.go
package main

import "bufio"
import "fmt"
import "os"
import "regexp"
import "strings"

func main() {

	if env_defined("RELAYCLIENT") ||
	   env_defined("TRUSTCLIENT") {
		fmt.Println()
		os.Exit(0)
	}

	file, err := os.Open("/var/qmail/control/badrcptto")
	if err != nil {
		fmt.Println()
		os.Exit(0)
	}
	defer file.Close()

	var recipient string = strings.ToLower(os.Getenv("SMTPRCPTTO"))

	if len(recipient) == 0 {
		fmt.Println()
		os.Exit(0)
	}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		match, _ := regexp.MatchString(recipient, scanner.Text())
		if match {
			fmt.Fprintf(os.Stderr, "%d Found %s in badrcptto list\n", os.Getppid(), recipient)
			fmt.Fprintf(os.Stdout, "E550 This address no longer accepts mail [%s]\n", recipient)
			os.Exit(0)
		}
	}

	fmt.Println()
	os.Exit(0)
}

func env_defined(key string) bool {
	value, exists := os.LookupEnv(key)
	_ = value

	return exists
}
